### PR TITLE
[Rest API] Change ``BrokersBase`` api ``getActiveBrokers`` and ``getLeaderBroker`` to pure async.

### PR DESCRIPTION
**Pull request by mattisonchao**
### Motivation

Async rest api is more efficient than sync. I think we could refactor all sync api to async.

### Modification

- Change ``getLeaderBroker`` from sync to async.
- Change ``getActiveBrokers`` from sync to async.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

- [x] `no-need-doc`

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java [hunk 1]
```diff
@@ -85,16 +85,21 @@
...
-            LOG.error("[{}] Failed to get active broker list: cluster={}", clientAppId(), cluster, e);
-            throw new RestException(e);
-        }
+    public void getActiveBrokers(@Suspended final AsyncResponse asyncResponse,
+                                 @PathParam("cluster") String cluster) {
+        validateSuperUserAccessAsync()
+                .thenCompose(__ -> validateClusterOwnershipAsync(cluster))
+                .thenCompose(__ -> pulsar().getLoadManager().get().getAvailableBrokersAsync())
+                .thenAccept(asyncResponse::resume)
+                .exceptionally(ex ->{
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java:94]:
```suggestion
                .exceptionally(ex -> {
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java [hunk 2]
```diff
@@ -85,16 +85,21 @@
...
+                    Throwable realCause = FutureUtil.unwrapCompletionException(ex);
+                    if (realCause instanceof WebApplicationException){
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java:96]:
```suggestion
                    if (realCause instanceof WebApplicationException) {
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java [hunk 3]
```diff
@@ -358,6 +375,25 @@ protected void validateClusterForTenant(String tenant, String cluster) {
         log.info("Successfully validated clusters on tenant [{}]", tenant);
     }
 
+    protected CompletableFuture<Void> validateClusterOwnershipAsync(String cluster){
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java:378]:
```suggestion
    protected CompletableFuture<Void> validateClusterOwnershipAsync(String cluster) {
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java [hunk 4]
```diff
@@ -358,6 +375,25 @@ protected void validateClusterForTenant(String tenant, String cluster) {
...
+        return getClusterDataIfDifferentCluster(pulsar(), cluster, clientAppId())
+                .thenAccept(differentClusterData -> {
+                    if (differentClusterData != null){
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java:381]:
```suggestion
                    if (differentClusterData != null) {
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java [hunk 5]
```diff
@@ -407,36 +432,28 @@ private URI getRedirectionUrl(ClusterData differentClusterData) throws Malformed
...
-            }
-        } else {
+                pulsar.getConfiguration().getClusterName().equals(cluster)) {
             clusterDataFuture.complete(null);
+            return clusterDataFuture;
         }
+        // redirect to the cluster requested
+        pulsar.getPulsarResources().getClusterResources().getClusterAsync(cluster)
+                .whenComplete((clusterDataResult, ex) -> {
+                    if (ex != null){
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java:445]:
```suggestion
                    if (ex != null) {
```

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java [hunk 6]
```diff
@@ -356,6 +357,21 @@ private void reapDeadBrokerPreallocations(List<String> aliveBrokers) {
...
+    @Override
+    public CompletableFuture<Set<String>> getAvailableBrokersAsync() {
+        CompletableFuture<Set<String>> getAvailableBrokersAsync = new CompletableFuture<>();
+        brokersData.listLocks(LoadManager.LOADBALANCE_BROKERS_ROOT)
+                .whenComplete((listLocks, ex) -> {
+                    if (ex != null){
+                        Throwable realCause = FutureUtil.unwrapCompletionException(ex);
+                        log.warn("Error when trying to get active brokers", realCause);
+                        getAvailableBrokersAsync.complete(loadData.getBrokerData().keySet());
+                    }
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java:369]: It's better to add `else` block

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/SimpleLoadManagerImpl.java [hunk 7]
```diff
@@ -303,6 +305,21 @@ public void disableBroker() throws Exception {
         return new HashSet<>(loadReports.listLocks(LOADBALANCE_BROKERS_ROOT).join());
     }
 
+    @Override
+    public CompletableFuture<Set<String>> getAvailableBrokersAsync() {
+        CompletableFuture<Set<String>> getAvailableBrokersAsync = new CompletableFuture<>();
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/SimpleLoadManagerImpl.java:310]: getAvailableBrokersAsync ->  future

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java (hunk 6 above)
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java:369]: getAvailableBrokersAsync -> future

Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java:371]: fixed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/SimpleLoadManagerImpl.java (hunk 7 above)
Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/SimpleLoadManagerImpl.java:310]: fixed

**Review by Technoboy- — APPROVED** [review]

**Comment by mattisonchao** [comment]
@Jason918  @liudezhi2098 PTAL :)

**Review by Jason918 — APPROVED** [review]

**Pull request merged**